### ISSUE TITLE
Function Calling Fine-tuning Support

### ISSUE THREAD

**Issue by jc-ryan** [huggingface/trl#2429]
Dear trl team,

As function calling fine-tuning becomes increasingly important for teaching language models to use tools effectively in agentic workflows, I'm pleased to see Hugging Face has unified the encoding and processing of various tool-calling messages through chat templates (https://huggingface.co/blog/unified-tool-use) .

I would like to inquire about trl's current or planned support for function calling fine-tuning. I haven't found related documentation in SFTTrainer (the conversation data format fields only include user and assistant roles). I believe supporting function calling would require at least these features:

1. Support for processing and encoding messages with user, assistant, and tool roles (already supported by the tokenizer library)
2. Loss calculation specifically for assistant parts (including both general messages and tool-calling messages)

Alternatively, is it possible to implement function calling fine-tuning using existing SFTTrainer features? For example, through a custom collate_fn similar to vlm_sft?

Thank you for your clarification!

**Comment by qgallouedec** [comment]
Thanks for suggesting this enhancement @jc-ryan. I agree it would be a great addition. cc @Rocketknight1

> 1. Support for processing and encoding messages with user, assistant, and tool roles (already supported by the tokenizer library)

It would probably be necessary to update the functions in `trl/data_utils.py` to include `tool` arg. This could be done as a first step.

> 2. Loss calculation specifically for assistant parts (including both general messages and tool-calling messages)

Can you elaborate on this part? Should the loss be calculated differently for the tool part?

> Alternatively, is it possible to implement function calling fine-tuning using existing SFTTrainer features? For example, through a custom collate_fn similar to vlm_sft?

I'm not sure I understand this point. With SFT at least, training is offline, so you shouldn't have to call the tools, right?

**Comment by jc-ryan** [comment]
> Loss calculation specifically for assistant parts (including both general messages and tool-calling messages)"

@qgallouedec Sorry for not being clear earlier. Regarding loss calculation, for tool using, there are several role messages including **system**, **user**, **assistant (tool_call)**, **tool (tool call results)**, and **assistant (plain text)**. We should mainly calculate the loss for the **assistant (tool_call)** and **assistant (plain text)** parts while masking the loss for other roles.

For the specific data format, you can refer to mistral: https://docs.mistral.ai/capabilities/finetuning/#2-function-calling-instruct

> Alternatively, is it possible to implement function calling fine-tuning using existing SFTTrainer features? For example, through a custom collate_fn similar to vlm_sft?

This part means I noticed that in the SFTTrainer documentation, you can write custom collate_fn for vlm_sft to support multimodal model input formats and loss masking (https://huggingface.co/docs/trl/sft_trainer#a-custom-collator-for-processing-multi-modal-data). So I'm not sure if this same approach can be applied to handle function calling data.

**Comment by qgallouedec** [comment]
Thanks for the clarification. Yes, you're right, the loss is calculated over the entire sequence by default (assistant and user). I think it's ok as a first approach, even if it's probably better to hide these parts in the loss.
For any contributor who wants to tackle this improvement, I suggest
1. update the functions in `trl/data_utils.py` to include tool arg.
2. update the trainers one by one to include tooling in their preprocessing stage (start with SFT)
3. find a clean way to hide the “tool” part of the assistant in the loss (data collator isn't IMHO). As a preliminary step, I'd also say prove that it's really worth doing.

**Comment by August-murr** [comment]
@qgallouedec 
Let's clarify before proceeding with the PR. 

given a dataset with three roles: user, assistant, and tool. The goal is to get the `SFTtrainer` and others to use a list of functions and feed that into the tokenizer (or processor). This will allow create a system prompt that tells the model which functions it has access to and how to use them.

and in `data_utils.py`, we’ll need to implement a function that checks whether the dataset uses tools. After that, we can modify the `SFTtrainer` and (other scripts) to preprocess all that before training.

**Comment by August-murr** [comment]
>Support for processing and encoding messages with user, assistant, and tool roles (already supported by the tokenizer library)

> 1. update the functions in `trl/data_utils.py` to include tool arg.
> 2. update the trainers one by one to include tooling in their preprocessing stage (start with SFT)

Here's something I don't get. given a conversational dataset with 3 roles of user, assistant and tool like [this dummy dataset](https://huggingface.co/datasets/August4293/tool_dummy_dataset) I made from the blog the tools dataset can be pre-processed like this:
```python
from datasets import load_dataset
dataset = load_dataset("August4293/tool_dummy_dataset")
tools = [get_current_temperature, get_current_time] #functions for the model to call
processed_text = processor.apply_chat_template(
    chat, 
    tools=tools
)
```
which results in:
```
<|begin_of_text|><|im_start|>system
You are a function calling AI model. You are provided with function signatures within <tools></tools> XML tags. You may call one or more functions to assist with the user query. Don't make assumptions about what values to plug into functions. Here are the available tools: <tools> {"type": "function", "function": {"name": "get_current_temperature", "description": "get_current_temperature(location: str) - Gets the temperature at a given location.

    Args:
        location(str): The location to get the temperature for, in the format "city, country"", "parameters": {"type": "object", "properties": {"location": {"type": "string", "description": "The location to get the temperature for, in the format \"city, country\""}}, "required": ["location"]}}
{"type": "function", "function": {"name": "get_current_time", "description": "get_current_time(location: str) - Gets the current time at a given location.

    Args:
        location(str): The location to get the time for, in the format "city, country"", "parameters": {"type": "object", "properties": {"location": {"type": "string", "description": "The location to get the time for, in the format \"city, country\""}}, "required": ["location"]}} </tools>Use the following pydantic model json schema for each tool call you will make: {"properties": {"name": {"title": "Name", "type": "string"}, "arguments": {"title": "Arguments", "type": "object"}}, "required": ["name", "arguments"], "title": "FunctionCall", "type": "object"}}
For each function call return a json object with function name and arguments within <tool_call></tool_call> XML tags as follows:
<tool_call>
{"name": <function-name>, "arguments": <args-dict>}
</tool_call><|im_end|>
<|im_start|>user
Hey, what's the weather like in Paris right now?<|im_end|>
<|im_start|>assistant
<tool_call>
{"name": "get_current_temperature", "arguments": {"location": "Paris, France"}}
</tool_call><|im_end|>
<|im_start|>tool
<tool_response>
22.0
</tool_response>
<|im_end|><|im_start|>assistant<|im_end|>
```

if it's already supported by the tokenizer library, why do we need to modify `SFTTrainer` to preprocess?
